### ISSUE TITLE
[plugin] download video

### ISSUE THREAD

**Issue by GreenLunar** [searx/searx#537]
Offer a plugin that, in addition to show/hide video, would provide a selection of available qualities of a given video to be downloaded.

**Comment by ghost** [comment]
That's an interesting idea. I am working on it here: https://github.com/ukwt/searx/tree/media-links

Example: https://i.imgur.com/uhejj1r.png

Only Oscar theme is supported at the moment.

I won't make pull request yet because it needs further testing and tinkering with the styles (notice how the download button looks).

There is one serious problem with this approach though: it works very slowly, extracting links for each result takes a few seconds, so the whole result page typically loads in 20-30 seconds (although this depends on the site, for example YouTube works much faster than Vimeo for me, so the whole page processing time becomes... bearable.)

This is unlikely to change, because most of that time is spent inside youtube-dl's code, and I am not sure there is any way to speed it up besides implementing a caching system.

**Comment by pointhi** [comment]
I would suggest to simply load the download urls when needed (clicking on the `download video` button causes an ajax call to searx, which then search for the video urls)

**Comment by GreenLunar** [comment]
> There is one serious problem with this approach though: it works very slowly, extracting links for each result takes a few seconds

I agree with @pointhi. In no way, this process should occur automatically to several results at once.

Suggestion: Once clicking on `download video` button, a popup with text "fetching list of video formats" with three squares (loading... indicator), and then the list of available video formats.

three squares - similarly to those displayed upon attempting to open map details.

**Comment by GreenLunar** [comment]
Suggestion: Once clicking on `download video` button, a popup with text "fetching list of video formats" with three squares (loading... indicator), and then the list of available video formats.

three squares - similarly to those displayed upon attempting to open map details.

**Comment by GreenLunar** [comment]
I guess, better than a popup.  Open a new element, similarly to the table opened upon opening map details.

**Comment by ghost** [comment]
Yup, that was not a good idea. Here is a second attempt: https://github.com/ukwt/searx/tree/video-links

Usually, results are sorted by the video quality in descending order, from best to worst. That's the reverse order of how youtube-dl returns them.

A few examples:

Download error:
http://i.imgur.com/OCAsTc8.png

Links were not found:
http://i.imgur.com/wGsl2O5.png

YouTube:
1. http://i.imgur.com/Ac5YQwJ.png
2. http://i.imgur.com/X72bN69.png
3. http://i.imgur.com/RvAYDBH.png

Vimeo:
1. http://i.imgur.com/C9sotk9.png
2. http://i.imgur.com/SFf0Icq.png

Dailymotion:
1. http://i.imgur.com/sTTFyun.png
2. http://i.imgur.com/tQ4G1QU.png

Any suggestions?

I think that table contains too much information, but condensing it into a small informative string could be problematic because field formats vary greatly across all supported engines (extractors, as youtube-dl calls them). Some extractors return resolution information, some do not; some return useful human-readable information in the "format" field, and some don't; some return not only video and/or audio tracks, but playlists also; the list goes on and on.

Also, I am not sure about formats columns ("audio" & "video"). Should I remove them?

**Comment by pointhi** [comment]
I would suggest adding some protection to this feature, so It can't be abused to easily

Simply hash the video-url with the secrete_key, and use it as additional request argument to get the download links of the result.

Like already done for the image proxy:
https://github.com/asciimoo/searx/blob/master/searx/webapp.py#L252
https://github.com/asciimoo/searx/blob/master/searx/webapp.py#L587

**Comment by ghost** [comment]
I was wondering what's the best way to prevent abusing the service. Thank you @pointhi, fixed.

**Comment by GreenLunar** [comment]
Well done!

> I think that table contains too much information, but condensing it into a small informative string could be problematic because field formats vary greatly across all supported engines.

I suggest to add an option to filter unwanted file formats (may be called `preferred media formats`), when there are several file formats to select from.

I am only downloading OGG (.oga or .ogv) or WebM when available.

**Comment by GreenLunar** [comment]
Well done!

> I think that table contains too much information, but condensing it into a small informative string could be problematic because field formats vary greatly across all supported engines.

I suggest to add an option to filter unwanted file formats (may be called `preferred media formats`), when there are several file formats to select from.

I am only downloading OGG (.oga or .ogv) or WebM when available.

**Comment by ghost** [comment]
@GreenLunar sure. I added settings for all file extensions supported by youtube-dl.

preferences page: http://i.imgur.com/BngbL7p.png

without filtering: http://i.imgur.com/1D2dCab.png

with 3gp filter: http://i.imgur.com/93r6LlD.png

with webm and mp4 filters: http://i.imgur.com/PU5d6Rg.png

Also I think it would be good to add an option to show all results without filtering when nothing of interest was found. I'll work on it in a little while.

**Comment by ghost** [comment]
OK, it's done.

Disable all formats:
http://i.imgur.com/ua2Ab6b.png

Search for something and click "download video":
http://i.imgur.com/RUegMzh.png

Click the link to view all results:
http://i.imgur.com/SLnVJHw.png

**Comment by pointhi** [comment]
I would display `nothing found`only when really nothing was found.

**Comment by ghost** [comment]
Makes sense. I updated my previous comment. Thank you @pointhi for valuable input.

**Comment by GreenLunar** [comment]
A suggestion for an improved UI.

Suppose I marked OGG as preferred format; results would be displayed as follows:

```
          OGG video 1
          OGG video 2
-------------------------------
   Show all available videos
```

By clicking on `Show all available videos`, user un-hides all the rest of the available formats to download.

**Comment by ghost** [comment]
Thanks @GreenLunar for the suggestion, that's how I should have done it in the first place. Fixed.

Here are the all possible scenarios:

Mobile view:
320x480:
http://i.imgur.com/WfUluOD.png
http://i.imgur.com/rRP7F89.png

360×640:
http://i.imgur.com/DtYmkAO.png
http://i.imgur.com/qSv8esI.png

768×1024:
http://i.imgur.com/qhLbIbC.png
http://i.imgur.com/AR6et9G.png

Nothing was found:
http://i.imgur.com/hSbFxZi.png

Nothing of interest was found:
http://i.imgur.com/EW55vAC.png
http://i.imgur.com/65XFCSm.png

Some results were filtered out:
http://i.imgur.com/dNWGtiV.png
http://i.imgur.com/uPyeMnv.png
http://i.imgur.com/RgWULMr.png

All extracted links are in some of the preferred formats:
http://i.imgur.com/7uQ4KIP.png

Connection error:
http://i.imgur.com/jqUzC3I.png

Any other ideas or suggestions? Table sorting wouldn't hurt, I guess.

**Comment by pointhi** [comment]
I would change in `Some results were filtered out` the text color of `Click here to display ... results` to mark that this is something to interact. (I think the css class `text-info` should be sufficient, normaly the color changes and a underline is displayed, when you hover such active elements on the oscar theme).

Also check how the table look like on small screens (mobile phone). (You can simply make the window smaller)

The rest looks fine for me.

**Comment by GreenLunar** [comment]
> Thanks @GreenLunar for the suggestion

You are welcome.

> that's how I should have done it in the first place.

I, too, did not think of it in the first place.  I guess you would have thought about this design by yourself, without me, sooner or later :)

> http://i.imgur.com/8kpmyPf.png

~~I prefer `No videos available to download` over `Nothing found`.~~
I prefer `No available media to download` over `Nothing found`.

> http://i.imgur.com/ED2BJJ9.png

`Click here to display all results` does not sound good to me.
Suggested text: `display all available results`.

> http://i.imgur.com/0U13LfR.png

`click here to display other NUMBER results` is better.
Perhaps `display other NUMBER results` (without `click here`) would be best.

**Comment by GreenLunar** [comment]
I forgot to point out that, it seems you have almost finished to work on this plugin :)

**Comment by ghost** [comment]
Thanks for your help everyone. I updated my previous comment.

**Comment by GreenLunar** [comment]
> Any other ideas or suggestions? Table sorting wouldn't hurt, I guess.

I guess so.

**Comment by GreenLunar** [comment]
(I will delete this comment)

> Yep, that's what I meant. Sorry if my message sounded a bit insulting, I didn't mean it that way.

No, no, no.

You wrote "that's how I should have done it in the first place" (not just you, but the both of us), and I agree, yet I presume that you would have thought about it too, sooner or later.

(I will delete this comment)

**Comment by ghost** [comment]
OK, finished. For example, here is a table sorted by the resolution column: http://i.imgur.com/Nhh5yKA.png

Any other ideas or suggestions?

**Comment by GreenLunar** [comment]
> Any other ideas or suggestions?

Filtering, like in spreadsheet editors; but this is for another time, I think.

**Comment by ghost** [comment]
@GreenLunar, done. Here is an example:

https://dl.dropboxusercontent.com/u/18321926/cast2.mkv

The implementation is rather naïve and inefficient though. It works fast for me, but could be better anyway. I'll have to optimize and prettify it a bit before making a pull request.

**Comment by GreenLunar** [comment]
@This is good.

I meant to a fixed filter; that is, if available extensions are `flv`, `ogg`, `webm`, then user would have three menu-item entries to filter upon clicking.

Is this plugin ready to be released?

**Comment by ghost** [comment]
> Is this plugin ready to be released?

Yeah, I think. There is one problem with some results though:

http://i.imgur.com/b0MqhX5.png

Take a look at table headers and notice how some icons are wrapped to the second line.

I've spent a few hours experimenting and haven't found a solution that would both keep icons on one line with text and keep them to the right of the cell, at least without major layout changes and changing the code. Many, many stackoverflow answers (among others) suggest using `white-space: nowrap`, which works by itself (that's how I 'fixed' this), but fails to work when icons are pulled to the side by `float: right`.

Does this look good enough?

http://i.imgur.com/vDLNRsO.png

If not, it would be great if someone more experienced with web design could suggest a solution.

**Comment by GreenLunar** [comment]
If you need someone to perform QA experiments, I am willing to assist.

> I've spent a few hours experimenting and haven't found a solution that would both keep icons on one line with text and keep them to the right of the cell, at least without major layout changes and changing the code.

I suggest not to include filter and sort buttons, for now.

> Yeah, I think. There is one problem with some results though:

I think we can waive some suggestions I have made over time, such as filtering and sorting; yet [trimming results](https://github.com/asciimoo/searx/issues/537#issuecomment-213254097) and a preference to set preferred formats resolutions etc. are more than enough; other enhancements may be implemented after the next release.

> Does this look good enough?

Yes.

**Comment by ghost** [comment]
Phew. I believe I've finally found a solution.

http://i.imgur.com/DDwbfGT.png

> I suggest not to include filter and sort buttons, for now.

Nah, it was almost ready, with the exception of those stubborn icons. But that seems to be fixed now.

I could probably have not wasted all that time re-implementing a (slower) wheel, but every third-party library I have tried either does not have all the functionality, or adds as many JS code as `searx` has already, if not more. I think that's too much for auxiliary functionality of an auxiliary plugin.

> If you need someone to perform QA experiments, I am willing to assist.

Sure, more testing on more systems and locales and whatnot is always good.

Downloading the code is easy:

```
git clone https://github.com/ukwt/searx.git
git checkout video-links
```

**Comment by GreenLunar** [comment]
> Nah, it was almost ready, with the exception of those stubborn icons. But that seems to be fixed now.

This plugin has more than I have expected; I think, even less is enough for initial release.

Yet, thank you for your efforts.  You have done a great work!

> Downloading the code is easy

I don't have the needed dependencies.

**Comment by GreenLunar** [comment]
This plugin has more than I have expected; I think, even less is enough for initial release.

Yet, thank you for your efforts.  You have done a great work!

**Comment by GreenLunar** [comment]
In the spirit of WebTorrent mentioned in #1908, we can load a [media detector/revealer](https://openuserjs.org/?q=video%20download) script to do that job on client-side.

**Comment by sjehuda** [comment]
Should we close this issue in favor of privacy respecting instances?
https://github.com/searxng/searxng/discussions/971

**Comment by unixfox** [comment]
> Should we close issue? [searxng/searxng#971](https://github.com/searxng/searxng/discussions/971)

What do you mean by that?

**Comment by sjehuda** [comment]
What I mean is that if you intend to use proxy services, then I assume this feature is going to be irrelevant.

**Comment by unixfox** [comment]
> What I mean is that if you intend to use proxy services, then I assume this feature is going to be irrelevant.

Okay, you mean people should use instead a service like invidious if they want to download a youtube video?

**Comment by sjehuda** [comment]
No.
If you ask me, I'll use youtube-dl.
But if Searx is to provide private instances, then a plugin would be needless.
That's my thought for this matter, but I'm not sure myself.

**Comment by sjehuda** [comment]
I was wrong.
A plugin as described here IS indeed relevant.

Please ignore my firstpost here.
Thank you
